tools/oeedger8r/Emitter.ml

```diff
@@ -217,102 +231,143 @@ let oe_get_param_size (ptype, decl, argstruct) =
 
 let oe_gen_wrapper_prototype (fd : func_decl) (is_ecall : bool) =
-  let plist_str = get_plist_str fd in
-  let retval_str =
-    if fd.rtype = Void then "" else sprintf "%s* _retval" (get_ret_tystr fd)
-  in
+  let args = List.map gen_parm_str fd.plist in
   let args =
-    if is_ecall then ["oe_enclave_t* enclave"; retval_str; plist_str]
-    else [retval_str; plist_str]
+    if fd.rtype <> Void then sprintf "%s* _retval" (get_tystr fd.rtype) :: args
+    else args
   in
-  let args = List.filter (fun s -> s <> "") args in
-  sprintf "oe_result_t %s(\n        %s)" fd.fname
-    (String.concat ",\n        " args)
-
-let emit_struct_or_union (os : out_channel) (s : struct_def) (union : bool) =
-  fprintf os "typedef %s %s {\n" (if union then "union" else "struct") s.sname ;
-  List.iter
-    (fun (atype, decl) ->
-      let dims = List.map (fun d -> sprintf "[%d]" d) decl.array_dims in
-      let dims_str = String.concat "" dims in
-      fprintf os "    %s %s%s;\n" (get_tystr atype) decl.identifier dims_str )
-    s.mlist ;
-  fprintf os "} %s;\n\n" s.sname
-
-let emit_enum (os : out_channel) (e : enum_def) =
-  let n = List.length e.enbody in
-  fprintf os "typedef enum %s {\n" e.enname ;
-  List.iteri
-    (fun idx (name, value) ->
-      fprintf os "    %s%s" name
-        ( match value with
-        | EnumVal (AString s) -> " = " ^ s
-        | EnumVal (ANumber n) -> " = " ^ string_of_int n
-        | EnumValNone -> "" ) ;
-      if idx != n - 1 then fprintf os ",\n" )
-    e.enbody ;
-  fprintf os "} %s;\n\n" e.enname
+  let args = if is_ecall then "oe_enclave_t* enclave" :: args else args in
+  sprintf "oe_result_t %s(%s)" fd.fname
+    ( if List.length args = 0 then ""
+    else if List.length args = 1 then List.hd args
+    else "\n    " ^ String.concat ",\n    " args )
 
 (** Emit [struct], [union], or [enum]. *)
-let emit_composite_type (os : out_channel) = function
-  | StructDef s -> emit_struct_or_union os s false
-  | UnionDef u -> emit_struct_or_union os u true
-  | EnumDef e -> emit_enum os e
+let emit_composite_type =
+  let emit_struct_or_union (s : struct_def) (union : bool) =
+    [ sprintf "typedef %s %s" (if union then "union" else "struct") s.sname
+    ; "{"
+    ; String.concat "\n"
+        (List.map
+           (fun (atype, decl) ->
+             let dims = List.map (fun d -> sprintf "[%d]" d) decl.array_dims in
+             let dims_str = String.concat "" dims in
+             sprintf "    %s %s%s;" (get_tystr atype) decl.identifier dims_str
+             )
+           s.mlist)
+    ; sprintf "} %s;\n" s.sname ]
+  in
+  let emit_enum (e : enum_def) =
+    [ sprintf "typedef enum %s" e.enname
+    ; "{"
+    ; String.concat ",\n"
+        (List.map
+           (fun (name, value) ->
+             sprintf "    %s%s" name
+               ( match value with
+               | EnumVal (AString s) -> " = " ^ s
+               | EnumVal (ANumber n) -> " = " ^ string_of_int n
+               | EnumValNone -> "" ) )
```

Comment:
Can you take a follow up item to see if we need to add a final enumeration member whose value equals to OE_ENUM_MAX?

Reply:
Filed #1767.